### ISSUE TITLE
Class method to get HTTP gem's User-Agent?

### ISSUE THREAD

**Issue by KelseyDH** [httprb/http#573]
I recently upgraded my `http` gem version recently, and we had lots of HTTP `webmock`s in our test break -- in part, because the `User-Agent` header of my http client had changed between versions.

Needing to manually update web mocks between HTTP requests is cumbersome, and what would be nice is if the `http` gem had a class level method like `HTTP.user_agent` one could call to programmatically get the `User-Agent` headers the current gem uses, so that the web mocks never break.

E.g. in my tests I have code like this:

```ruby
  def stub_confirmations_api(response_code)
    stub_request(:post, "#{my_url}/confirmations").
      with(
        headers: {
          "Auth-Id": ENV.fetch("AUTH_ID"),
          "Auth-Token": ENV.fetch("AUTH_TOKEN"),
          "Connection": 'close',
          "Host": 'localhost',
          "User-Agent": 'http.rb/4.1.1'
        }
      ).to_return(status: response_code, body: nil, headers: {})
  end
```

That needed to be manually updated to:

```ruby
  def stub_confirmations_api(response_code)
    stub_request(:post, "#{my_url}/confirmations").
      with(
        headers: {
          "Auth-Id": ENV.fetch("AUTH_ID"),
          "Auth-Token": ENV.fetch("AUTH_TOKEN"),
          "Connection": 'close',
          "Host": 'localhost',
          "User-Agent": 'http.rb/4.2.0'
        }
      ).to_return(status: response_code, body: nil, headers: {})
  end
```

This maintenance burden could be avoided if for `User-Agent` I could just invoke a method from the gem, to give me its current user-agent (e.g. `HTTP.user_agent`).  E.g.:

```ruby
  def stub_confirmations_api(response_code)
    stub_request(:post, "#{my_url}/confirmations").
      with(
        headers: {
          "Auth-Id": ENV.fetch("AUTH_ID"),
          "Auth-Token": ENV.fetch("AUTH_TOKEN"),
          "Connection": 'close',
          "Host": 'localhost',
          "User-Agent": HTTP.user_agent
        }
      ).to_return(status: response_code, body: nil, headers: {})
  end
```

This would prevent upgrades of the gem from breaking webmocks for no other reason than the fact that the gem's version was updated.

**Comment by tarcieri** [comment]
instead of:

```ruby
"User-Agent": 'http.rb/4.2.0'
```

you can do:

```ruby
"User-Agent": "http.rb/#{HTTP::VERSION}"
```

There is no expectation of ever changing this format. I agree a method would be nice, but the above works today and is unlikely to ever break in the future.

**Comment by mikegee** [comment]
👍 

> Having a method to be able to set or get `HTTP.default_headers` as a hash would also be nice.

I agree with the sentiment in this request. In my apps, I have a method to get a new client that always sets the user-agent. Here's an example:

```ruby
    def http_client
      HTTP
        .headers('User-Agent' => "MyAppName #{VERSION}",
                 'Accept'     => 'application/json')
        .timeout(:global, connect: TIMEOUT, write: TIMEOUT, read: TIMEOUT)
    end
```

I don't think it's appropriate for requests from my application out to the wider internet to have a user-agent from the underlying library.

**Comment by ixti** [comment]
For simplicity of those who need to fetch default user agent, we should expose it as a constant IMO:

``` ruby
class HTTP::Client
  DEFAULT_USER_AGENT = "http.rb/#{HTTP::VERSION}"
end

**Comment by KelseyDH** [comment]
> For simplicity of those who need to fetch default user agent, we should expose it as a constant IMO:
> 
> ```ruby
> class HTTP::Client
>   DEFAULT_USER_AGENT = "http.rb/#{HTTP::VERSION}"
> end
> ```

I like this idea.  Something like `HTTP::DEFAULT_USER_AGENT` would be concise, yet clear enough that those modifying their `User-Agent` headers would know it doesn't apply to them.

**Comment by tarcieri** [comment]
Placing it in `HTTP::Client` probably makes more sense as the `HTTP` namespace tends to get a bit polluted

**Comment by KelseyDH** [comment]
> Placing it in `HTTP::Client` probably makes more sense as the `HTTP` namespace tends to get a bit polluted

True, `HTTP` is a common namespace.  `HTTP::Client::DEFAULT_USER_AGENT` works.

**Comment by ixti** [comment]
`mattr_accessor` is a Rails method. And I don't think it's correct to override user agent of default client.

**Comment by ixti** [comment]
Also, I feel myself a bit stupid, but default user agent is long-time available as `HTTP::Request::USER_AGENT`

**Comment by ixti** [comment]
Probably worth to move it to `HTTP::Client` though to make it more obvious.

**Comment by sfisher** [comment]
The easy solution to this problem is to either leave off the user-agent matching requirement for your webmock or else replace it with a regular expression matcher that matches all versions of http.rb user agents.

I had this problem before and it was because I was a lazy programmer who just copied and pasted the suggested webmock when an rspec test failed. You can make webmocks as specific or broad as you want and it's somewhat well documented online or in the github Readme. It seems like people are going to a lot of work to put a specific version into their tests when there is no need to. If you don't care that http.rb is a specific version in your tests then leave the user-agent condition out or make it more broad with a regular expression.

**Comment by sferik** [comment]
In that case, I’m going to close this issue. We can reevaluate for v7.0.0. 😉
